### ISSUE TITLE
Notebook codemirror Emacs keybindings error in IPy 3.0/4.0

### ISSUE THREAD

**Issue by histed** [ipython/ipython#8642]
In IPython 2.1.0, using codemirror to set Emacs keybindings worked for me. 
In IPython 3.2.0 that stopped working.  I copied codemirror's emacs.js from my 2.1.0 to my 3.0 install and it seems to have fixed the problem.  Problem with emacs.js in IPy 3.0?

I originally followed the instructions here:
https://github.com/ipython/ipython/issues/2319
(although the notebook referenced there now is a broken link, and I couldn't find these docs anywhere else on the web.)

My custom.js:
```
function to(mode) {
    var mode = mode || 'vim'
    function to_mode(c) { return c.code_mirror.setOption('keyMap', mode);};
    IPython.notebook.get_cells().map(to_mode);
    IPython.Cell.options_default.cm_config.keyMap = mode;
}
require(["base/js/events"], function (events) {
    $([IPython.events]).on("app_initialized.NotebookApp", function () {
        require(["components/codemirror/keymap/emacs"], function () {
            to('emacs')
            console.log('emacs.js loaded')
        });
    });
});
```

emacs.js is at lib/python2.7/site-packages/IPython/html/static/components/codemirror/keymap/emacs.js.

With the 3.0 emacs.js, I got this error on the Chrome JS console:
```
require.js?v=640929dac3c23a448d2eebc37bc32062:166 
Uncaught Error: Load timeout for modules: components/codemirror/lib/codemirror
http://requirejs.org/docs/errors.html#timeout
```
I didn't see any other errors, but perhaps require.js was suppressing them, as described at
http://requirejs.org/docs/errors.html#timeout

**Comment by aiguofer** [comment]
Thanks! I've been struggling with this and it works with the 2.10 files (https://raw.githubusercontent.com/ipython/ipython-components/2.1.0/codemirror/keymap/emacs.js).

It looks like they changed the way that CodeMirror loads modules, the new version has this at the top:
```
(function(mod) {
  if (typeof exports == "object" && typeof module == "object") // CommonJS
    mod(require("../lib/codemirror"));
  else if (typeof define == "function" && define.amd) // AMD
    define(["../lib/codemirror"], mod);
  else // Plain browser env
    mod(CodeMirror);
})(function(CodeMirror) {
```
I think now CodeMirror needs to be informed of which modules to load instead of using require to load manually. However, I'm not familiar enough with this and this workaround seems to work for now.

**Comment by histed** [comment]
Note for ipython 4.0, custom.js is in $HOME/.jupyter/custom and the following lines work:
```
function to(mode) {
    var mode = mode || 'vim'
    function to_mode(c) { return c.code_mirror.setOption('keyMap', mode);};
    IPython.notebook.get_cells().map(to_mode);
    IPython.Cell.options_default.cm_config.keyMap = mode;
}
require(["base/js/namespace", "base/js/events"], function (IPython, events) {
    events.on("notebook_loaded.Notebook", function () {
        require(["components/codemirror/keymap/emacs"], function () { 
            to('emacs')
            console.log('emacs.js loaded')
        });
    });
});
```
I don't understand why but waiting on app_initialized.NotebookApp no longer works; notebook_loaded.Notebook does.

**Comment by dsblank** [comment]
Interesting find regarding the events to firing correctly. Maybe that is what I am seeing here:

https://github.com/jupyter/notebook/issues/412

**Comment by Carreau** [comment]
Grumble grumble grumble Javascript.

**Comment by Carreau** [comment]
Ok, we cannot minify/bundle Codemirror  as mode do require `../lib/codemirror` that for requirejs resolve differently that `compents/codemirror/lib/codemirror` id codemirror is bundled. 

Which lead to things like emacs mode that do:

```
CodeMirror = require('../lib/codemirror')
Codemirror.keyMap.emacs = ...
```


To not work as `Codemirror != window.CodeMirror`

@minrk  ?

**Comment by Carreau** [comment]
Hum, I might have a solution, which I'm not sure why it works. 
Can you try replacing `"components/codemirror/keymap/emacs"` by `"codemirror/keymap/emacs"` (no leading `compoents`)

**Comment by histed** [comment]
@Carreau thanks for looking into this.  I dropped my custom.js here to help people looking for emacs bindings in ipy 4.0/jupyter.  It will be hard for me to find the time to do much testing over the next few weeks.

**Comment by tbekolay** [comment]
Following this thread, I added the snippet in the original post at `~/.jupyter/custom/custom.js`, replaced `"components/codemirror/keymap/emacs"` with `"codemirror/keymap/emacs"` and emacs keybindings now work for me.

I get the following warning in the console: `"accessing Cell is deprecated. Use require("notebook/js/cell").Cell"` but it doesn't affect anything. Presumably the snippet should be updated accordingly though...

**Comment by Carreau** [comment]
Can you post your final version ? If I have some time I'll have a look.

**Comment by deuscovrigus** [comment]
@histed  It looks like that the  require(["base/js/events"], loads only after the notebook app has initialized on Firefox, Ubuntu 14.04. Ipython 4.0.0. My final snippet is

```javascript
  $([IPython.events]).on("app_initialized.NotebookApp", function () {
          require(["codemirror/keymap/emacs"], function () {
            to('emacs')
            console.log('emacs.js loaded')
        });
    });
```

**Comment by sashkab** [comment]
Thank you for the hints, I finally was able to update my custom.js. Here is my final version, if anyone cares. OSX 10.10.5, IPython 4.0.0:

```js
require([
    'base/js/namespace',
    'base/js/events',
], function (IPython, events) {

    function to(mode) {
        function to_mode(c) { 
            return c.code_mirror.setOption('keyMap', mode);
        };
        IPython.notebook.get_cells().map(to_mode);
        require("notebook/js/cell").Cell.options_default.cm_config.keyMap = mode;
    };

    events.on('notebook_loaded.Notebook', function () {
        require(["codemirror/keymap/emacs"], function () {
            to('emacs')
            console.log('emacs.js loaded')
        });
    });
});
```

**Comment by deuscovrigus** [comment]
Yep yours works for me too @sashkab

**Comment by tbekolay** [comment]
@sashkab's snippet works perfectly for me as well! Gets rid of the deprecation warning as well.

**Comment by aiguofer** [comment]
+1 :)

**Comment by geraldrude** [comment]
I can't thank you enough for this solution!  I was encouraged long ago to use emacs as my text editor, and this decision wired my brain to use ctrl key text navigation. I can now go back to using jupyter without having to pick up my mouse every 30 seconds.  Thank you very much!

**Comment by nrhine1** [comment]
Hope this helps someone: if you want to do system paste with the standard Ctrl-v, this slight modification (` mode.setOption("extraKeys", {"Ctrl-v": false});`) seems to enable it by turning off the default Ctrl-v binding. It would be nice to have a way of loading `.emacs` files but I haven't really done my research, perhaps someone already has a solution.

```javascript
require([
    'base/js/namespace',
    'base/js/events',
], function (IPython, events) {

    function to(mode) {
	mode.setOption("extraKeys", {"Ctrl-v": false});

        function to_mode(c) { 
            return c.code_mirror.setOption('keyMap', mode);
        };
        IPython.notebook.get_cells().map(to_mode);
        require("notebook/js/cell").Cell.options_default.cm_config.keyMap = mode;
	
    };

    events.on('notebook_loaded.Notebook', function () {
        require(["codemirror/keymap/emacs"], function () {
            to('emacs')
            console.log('emacs.js loaded')
        });
    });
});```

**Comment by tbekolay** [comment]
@nrhine1 This doesn't work for me. If you wanted to set this option, you would have to do it like so:

```javascript
require([
    'base/js/namespace',
    'base/js/events',
], function (IPython, events) {

    function to(mode) {
        // Note: This doesn't work because 'mode' is a string
        // mode.setOption("extraKeys", {"Ctrl-v": false});
        function to_mode(c) { 
            c.code_mirror.setOption('keyMap', mode);
            c.code_mirror.getOption('extraKeys')["Ctrl-V"] = false;  // added
        };
        IPython.notebook.get_cells().map(to_mode);
        require("notebook/js/cell").Cell.options_default.cm_config.keyMap = mode;
        require("notebook/js/cell").Cell.options_default.cm_config.extraKeys["Ctrl-V"] = false;  // added
    };

    events.on('notebook_loaded.Notebook', function () {
        require(["codemirror/keymap/emacs"], function () {
            to('emacs')
            console.log('emacs.js loaded')
        });
    });
});
```

Note the two lines with the `// added` comment. However, this doesn't work for me; I still get the cursor movement behavior. It may have appeared to work for you because your snippet raises an error, which disables the Emacs keybindings entirely.

I tried a bunch of things (changing `extraKeys`, `customKeys`, etc) but nothing that I could do would restore the default `Ctrl-V` pasting behavior. I think it's just not possible with CodeMirror at the moment! Also, loading `.emacs` files definitely is not possible with CodeMirror, and likely never will be; the architectures are entirely different.

----

One annoyance I was having is that `Ctrl-Shift--`, which I use as undo, is mapped by Jupyter as "Split cell at cursor". I have modified my `jupyter.js` snippet to remove this shortcut (not sure if this is the right place for it, but it works.

```javascript
require([
    'base/js/namespace',
    'base/js/events',
], function (IPython, events) {

    function to(mode) {
        function to_mode(c) {
            c.code_mirror.setOption('keyMap', mode);
        };
        IPython.notebook.get_cells().map(to_mode);
        require("notebook/js/cell").Cell.options_default.cm_config.keyMap = mode;
    };

    events.on('notebook_loaded.Notebook', function () {
        require(["codemirror/keymap/emacs"], function () {
            to('emacs');
            console.log('emacs.js loaded');
        });

        // Also remove Ctrl-Shift-- as Emacs uses this for undo
        IPython.keyboard_manager.edit_shortcuts.remove_shortcut('ctrl-shift--');
        IPython.keyboard_manager.edit_shortcuts.remove_shortcut('ctrl-shift-subtract');
    });
});
```

Note the last three lines in the `event.on` function.

**Comment by nrhine1** [comment]
@tbekolay  you're right. It was causing an error, which prevented the modification `Ctrl-v` to `emacs` mode. It is obviously a string, somehow my wealth of JS experience led me astray ;)

However, I was able to get it to work (both emacs mode and changing `Ctrl-V` functionality) by using `setOption('extraKeys')`

```javascript
require([
    'base/js/namespace',
    'base/js/events',
], function (IPython, events) {

    function to(mode) {


        function to_mode(c) { 
            c.code_mirror.setOption('keyMap', mode);
            c.code_mirror.setOption('extraKeys')["Ctrl-v"] = false;  // modified
        };
        IPython.notebook.get_cells().map(to_mode);	
        require("notebook/js/cell").Cell.options_default.cm_config.keyMap = mode;
    };

    events.on('notebook_loaded.Notebook', function () {
        require(["codemirror/keymap/emacs"], function () {
            to('emacs')
            console.log('emacs.js loaded')
        });
    });
});
```

On the `.emacs` loading -- without knowing the specifics, I think it would be possible for an ipython install to request access to the `.emacs` file and parse only the `global-set-key` calls, which would be retrieved and set as `extraKeys`, but this would definitely require a bit of engineering. It is probably infeasible to load the entire Emacs OS into ipython, so modification of `custom.js` seems to be the good enough solution for now.

**Comment by nrhine1** [comment]
@tbekolay you're right, it was because `emacs` mode failed to load at all. My wealth of JS experience seemed to have failed me... Thanks for trying it out and letting me know!

However, I got it to work (both `emacs` mode and `Ctrl-V` copy from system buffer;) by using:

```javascript
c.code_mirror.setOption('keyMap', mode);
c.code_mirror.setOption('extraKeys', {"Ctrl-V" : false});  // added
```

so the total snippet becomes:

```javascript
require([
    'base/js/namespace',
    'base/js/events',
], function (IPython, events) {

    function to(mode) {


        function to_mode(c) { 
            c.code_mirror.setOption('keyMap', mode);
            c.code_mirror.setOption('extraKeys', {"Ctrl-V" : false});  // added
        };
        IPython.notebook.get_cells().map(to_mode);	
        require("notebook/js/cell").Cell.options_default.cm_config.keyMap = mode;
    };

    events.on('notebook_loaded.Notebook', function () {
        require(["codemirror/keymap/emacs"], function () {
            to('emacs')
            console.log('emacs.js loaded')
        });
    });
});
```

Hopefully this works for you too? I'm using ipython notebook version 3.2.1 on Ubuntu 14.04.

**Comment by takluyver** [comment]
Closing as it looks like this got more or less worked out. If there's still anything to do, please open a new issue on the jupyter/notebook repo.
